iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Client.Exceptions;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.Samples
+{
+    public class DeviceReconnectionSample
+    {
+        private static readonly Random s_randomGenerator = new Random();
+        private const int TemperatureThreshold = 30;
+        private static readonly TimeSpan s_sleepDuration = TimeSpan.FromSeconds(5);
+        private static readonly TimeSpan s_operationTimeout = TimeSpan.FromHours(1);
+
+        private readonly string _deviceConnectionString;
+        private readonly TransportType _transportType;
+        private readonly ILogger _logger;
+        private readonly object _lock = new object();
+
+        private static DeviceClient s_deviceClient;
+        private static ConnectionStatus connectionStatus;
+        private static bool wasConnected;
+
+        public DeviceReconnectionSample(string deviceConnectionString, TransportType transportType, ILogger logger)
+        {
+            _logger = logger;
+            _deviceConnectionString = deviceConnectionString;
+            _transportType = transportType;
+
+            InitializeClient();
+        }
+
+        private void ConnectionStatusChangeHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
+        {
+            _logger.LogDebug($"Connection status changed: status={status}, reason={reason}");
+
+            connectionStatus = status;
+            switch (connectionStatus)
+            {
+                case ConnectionStatus.Connected:
+                    _logger.LogDebug("### The DeviceClient is CONNECTED; all operations will be carried out as normal.");
+
+                    wasConnected = true;
```

Comment:
isConnected?

Reply:
Actually this variable is supposed to indicate if the client was ever connected. This is because, if the initial open async fails, then the client reports a Disconnected status, but the underneath layer isn't initialized yet. So when we do a dispose and re-open, it is messing up the internal state, and throws an object disposed exception.
What I can gather from here is, if I call dispose on a client that doesn't have the underlying layer initialized, it is getting messed up.

Now that I think about it more, let me run some tests on the code to see if I can root cause it better. 
This might be a fix only for the symptom, but not the root cause.

Reply:
So `wasEverConnected`?

Reply:
sure, sounds better